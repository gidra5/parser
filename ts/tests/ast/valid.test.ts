import { describe, expect } from "vitest";
import { it, fc, test } from "@fast-check/vitest";
import { Iterator } from "iterator-js";
import { infixArithmeticOps, prefixArithmeticOps } from "../../src/parser";
import {
  group,
  infix,
  name,
  number,
  prefix,
  string,
} from "../../src/parser/ast";
import { matchSeparators } from "../../src/parser/utils";
import { treeTestCase, treeTestCaseArgs } from "./utils";

describe("comments", () => {
  test("comment", () => {
    const src = `// comment\n123`;
    treeTestCase(src);
  });

  test("comment block", () => {
    const src = `/* comment block */123`;
    treeTestCase(src);
  });
});

describe("expressions", () => {
  describe("values", () => {
    test("number", () => {
      const src = `123`;
      treeTestCase(src);
    });

    test("string", () => {
      const src = `"string"`;
      treeTestCase(src);
    });

    test("true", () => {
      const src = `true`;
      treeTestCase(src);
    });

    test("false", () => {
      const src = `false`;
      treeTestCase(src);
    });
  });

  describe("fixity expressions", () => {
    test("name", () => {
      const src = `name`;
      treeTestCase(src);
    });

    test.todo("operator", () => {
      const src = `+`;
    });

    test("group", () => {
      const src = `(1)`;
      treeTestCase(src);
    });

    test("prefix", () => {
      const src = `+123`;
      treeTestCase(src);
    });

    test("postfix", () => {
      const src = `123--`;
      treeTestCase(src);
    });

    test("infix", () => {
      const src = `123+456`;
      treeTestCase(src);
    });

    test("mixfix", () => {
      const src = `123 < 456 < 789`;
      treeTestCase(src);
    });
  });

  describe("arithmetics", () => {
    for (const [opName, op] of infixArithmeticOps) {
      test(opName, () => {
        const src = `123 ${op} 456`;
        treeTestCase(src);
      });
    }

    for (const [opName, op] of prefixArithmeticOps) {
      test(opName, () => {
        const src = `${op}123`;
        treeTestCase(src);
      });
    }

    test("postfix decrement", () => {
      const src = `123--`;
      treeTestCase(src);
    });

    test("postfix increment", () => {
      const src = `123++`;
      treeTestCase(src);
    });

    it(...treeTestCaseArgs("1 + 2^3 * 4"));
    it(...treeTestCaseArgs("--i"));
    it(...treeTestCaseArgs("++i"));
    it(...treeTestCaseArgs("-(a+b)"));

    it(
      ...treeTestCaseArgs(
        "(2^2-5+7)-(-i)+ (j)/0 - 1*(1*f)+(27-x )/q + send(-(2+7)/A,j, i, 127.0 ) + 1/1"
      )
    );
  });

  describe("boolean expressions", () => {
    test("in", () => {
      const src = `123 in 456`;
      treeTestCase(src);
    });

    test("is", () => {
      const src = `123 is 456`;
      treeTestCase(src);
    });

    test("and", () => {
      const src = `123 and 456`;
      treeTestCase(src);
    });

    test("or", () => {
      const src = `123 or 456`;
      treeTestCase(src);
    });

    test("not", () => {
      const src = `not 123`;
      treeTestCase(src);
    });

    test("equal", () => {
      const src = `123 == 456`;
      treeTestCase(src);
    });

    test("not equal", () => {
      const src = `123 != 456`;
      treeTestCase(src);
    });

    describe("comparators", () => {
      const comparators = [
        ["<", "<="],
        [">", ">="],
      ];

      for (const op of Iterator.iter(comparators).flat()) {
        test(`comparator ${op}`, () => {
          const src = `123 ${op} 456`;
          treeTestCase(src);
        });
      }

      for (const [op1, op2] of Iterator.iter(comparators).flatMap((pair) =>
        Iterator.iter(pair).power(2)
      )) {
        test(`range ${op1} ${op2}`, () => {
          const src = `123 ${op1} x ${op2} 456`;
          treeTestCase(src);
        });
      }
    });
  });

  describe("function expressions", () => {
    test("funciton multiple params", () => {
      const src = `fn x y -> x + y`;
      treeTestCase(src);
    });

    test("function", () => {
      const src = `x -> x`;
      treeTestCase(src);
    });

    describe("application", () => {
      test("function call", () => {
        const src = `f x`;
        treeTestCase(src);
      });

      test("function call multiple args", () => {
        const src = `f x y`;
        treeTestCase(src);
      });

      it(
        ...treeTestCaseArgs("send((1+2), 3)", undefined, {
          send: {
            separators: matchSeparators(["send"]),
            precedence: [null, Infinity],
          },
        })
      );

      it(...treeTestCaseArgs("send(2, 3)"));
      it(...treeTestCaseArgs("(send)(2, 3)"));
      it(...treeTestCaseArgs("(send 1)(2, 3)"));
      it(...treeTestCaseArgs("(send 1 2)(2, 3)"));
      it(...treeTestCaseArgs("send 1 + 2"));
      it(...treeTestCaseArgs("send 1 (2, 3)"));
      it(...treeTestCaseArgs("send a (2, 3)"));
      it(...treeTestCaseArgs("a + send (2, 3)"));
      it(...treeTestCaseArgs("a + send 1 + 2"));
    });
  });

  describe("pattern matching", () => {
    test.todo("match", () => {
      const src = `match x { 1 -> 2; 3 -> 4 }`;
    });

    test.todo("in function parameters", () => {
      const src = `(x, y) -> x + y`;
    });

    test.todo("with 'is' operator", () => {
      const src = `x is (a, b)`;
    });

    test.todo("with placeholder", () => {
      const src = `x is (_, b)`;
    });
  });

  describe("structured programming", () => {
    test.todo("if-then", () => {
      const src = `if true: 123`;
    });

    test.todo("if-then-else", () => {
      const src = `if true: 123 else 456`;
    });
    test.todo("block", () => {
      const src = `{ 123 }`;
    });

    test.todo("for loop", () => {
      const src = `for x in [1, 2, 3]: x`;
    });

    test.todo("while loop", () => {
      const src = `while true: 123`;
    });

    test.todo("while loop break", () => {
      const src = `while true: break 1`;
    });

    test.todo("while loop continue", () => {
      const src = `while true: continue`;
    });

    test.todo("block break", () => {
      const src = `{ break 1 }`;
    });

    test.todo("labeled expression", () => {
      const src = `label: 123`;
    });

    test.todo("expression-label", () => {
      const src = `[123]: 456`;
    });

    test.todo("return", () => {
      const src = `() => { return 123 }`;
    });

    test.todo("block variable declaration", () => {
      const src = `{ x := 123 }`;
    });

    test.todo("block variable declaration with type", () => {
      const src = `{ x: number := 123 }`;
    });

    test.todo("block variable assignment", () => {
      const src = `{ x = 123 }`;
    });

    test.todo("block pattern matching", () => {
      const src = `{ x, y = 123, 456 }`;
    });
  });

  describe("data structures", () => {
    it(...treeTestCaseArgs("(-(2+7)/A,j, i, 127.0 )"));

    test("unit", () => {
      const src = `()`;
      treeTestCase(src);
    });

    test("tuple", () => {
      const src = `1, 2`;
      treeTestCase(src);
    });

    test.todo("list", () => {
      const src = `[1, 2]`;
    });

    test.todo("record", () => {
      const src = `record { a: 1, b: 2 }`;
    });

    test.todo("set", () => {
      const src = `set { 1, 2 }`;
    });

    test.todo("map", () => {
      const src = `map { 1: 2, 3: 4 }`;
    });

    test.todo("field access", () => {
      const src = `x.y`;
    });

    test.todo("field access dynamic", () => {
      const src = `x[y]`;
    });
  });
});

describe("programs", () => {
  describe("script", () => {
    test.todo("variable", () => {
      const src = `x := 123`;
    });
    test.todo("import", () => {
      const src = `x := 123;  import "a" as b`;
    });
    test.todo("import with", () => {
      const src = `x := 123;  import "a" as b with x `;
    });
    test.todo("export", () => {
      const src = `x := 123; export x`;
    });
    test.todo("export as", () => {
      const src = `x := 123; export x as y`;
    });
  });

  describe("module", () => {
    test.todo("import/use", () => {
      const src = `use "a" as b`;
    });
    test.todo("import/use with", () => {
      const src = `use "a" as b with c`;
    });
    test.todo("import/use with external", () => {
      const src = `use "a" as b with external c`;
    });
    test.todo("external", () => {
      const src = `external y`;
    });
    test.todo("private declare", () => {
      const src = `z := y+1`;
    });
    test.todo("public declare", () => {
      const src = `export x := z+123`;
    });
    test.todo("export main", () => {
      const src = `export (args) -> {}`;
    });
  });
});
