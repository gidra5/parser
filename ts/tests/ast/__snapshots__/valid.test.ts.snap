// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comments > comment 1`] = `
{
  "children": [],
  "name": "number",
  "value": 123,
}
`;

exports[`comments > comment block 1`] = `
{
  "children": [],
  "name": "number",
  "value": 123,
}
`;

exports[`expressions > arithmetics > add 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > decrement 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "prefixDecrement",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > divide 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "/",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > increment 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "prefixIncrement",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > modulo 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "%",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > multiply 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "*",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > negate 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "negate",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > postfix decrement 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "postfixDecrement",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "postfix",
}
`;

exports[`expressions > arithmetics > postfix increment 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "postfixIncrement",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "postfix",
}
`;

exports[`expressions > arithmetics > power 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "^",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > produces correct tree for '(2^2-5+7)-(-i)+ (j)/0 - 1*(1*f)+(27-x )/q + send(-(2+7)/A,j, i, 127.0 ) + 1/1' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "+",
        },
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "+",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "group",
                  "value": "-",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "name": "group",
                      "value": "+",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "name": "group",
                          "value": "-",
                        },
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [],
                                  "name": "group",
                                  "value": "+",
                                },
                                {
                                  "children": [
                                    {
                                      "children": [],
                                      "name": "group",
                                      "value": "-",
                                    },
                                    {
                                      "children": [
                                        {
                                          "children": [],
                                          "name": "group",
                                          "value": "^",
                                        },
                                        {
                                          "children": [],
                                          "name": "number",
                                          "value": 2,
                                        },
                                        {
                                          "children": [],
                                          "name": "number",
                                          "value": 2,
                                        },
                                      ],
                                      "name": "infix",
                                    },
                                    {
                                      "children": [],
                                      "name": "number",
                                      "value": 5,
                                    },
                                  ],
                                  "name": "infix",
                                },
                                {
                                  "children": [],
                                  "name": "number",
                                  "value": 7,
                                },
                              ],
                              "name": "infix",
                            },
                          ],
                          "name": "group",
                          "value": "parens",
                        },
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [],
                                  "name": "group",
                                  "value": "negate",
                                },
                                {
                                  "children": [],
                                  "name": "name",
                                  "value": "i",
                                },
                              ],
                              "name": "prefix",
                            },
                          ],
                          "name": "group",
                          "value": "parens",
                        },
                      ],
                      "name": "infix",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "name": "group",
                          "value": "/",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "name": "name",
                              "value": "j",
                            },
                          ],
                          "name": "group",
                          "value": "parens",
                        },
                        {
                          "children": [],
                          "name": "number",
                          "value": 0,
                        },
                      ],
                      "name": "infix",
                    },
                  ],
                  "name": "infix",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "name": "group",
                      "value": "*",
                    },
                    {
                      "children": [],
                      "name": "number",
                      "value": 1,
                    },
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [],
                              "name": "group",
                              "value": "*",
                            },
                            {
                              "children": [],
                              "name": "number",
                              "value": 1,
                            },
                            {
                              "children": [],
                              "name": "name",
                              "value": "f",
                            },
                          ],
                          "name": "infix",
                        },
                      ],
                      "name": "group",
                      "value": "parens",
                    },
                  ],
                  "name": "infix",
                },
              ],
              "name": "infix",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "group",
                  "value": "/",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "name": "group",
                          "value": "-",
                        },
                        {
                          "children": [],
                          "name": "number",
                          "value": 27,
                        },
                        {
                          "children": [],
                          "name": "name",
                          "value": "x",
                        },
                      ],
                      "name": "infix",
                    },
                  ],
                  "name": "group",
                  "value": "parens",
                },
                {
                  "children": [],
                  "name": "name",
                  "value": "q",
                },
              ],
              "name": "infix",
            },
          ],
          "name": "infix",
        },
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "application",
            },
            {
              "children": [],
              "name": "name",
              "value": "send",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "name": "group",
                      "value": ",",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "name": "group",
                          "value": ",",
                        },
                        {
                          "children": [
                            {
                              "children": [],
                              "name": "group",
                              "value": ",",
                            },
                            {
                              "children": [
                                {
                                  "children": [],
                                  "name": "group",
                                  "value": "/",
                                },
                                {
                                  "children": [
                                    {
                                      "children": [],
                                      "name": "group",
                                      "value": "negate",
                                    },
                                    {
                                      "children": [
                                        {
                                          "children": [
                                            {
                                              "children": [],
                                              "name": "group",
                                              "value": "+",
                                            },
                                            {
                                              "children": [],
                                              "name": "number",
                                              "value": 2,
                                            },
                                            {
                                              "children": [],
                                              "name": "number",
                                              "value": 7,
                                            },
                                          ],
                                          "name": "infix",
                                        },
                                      ],
                                      "name": "group",
                                      "value": "parens",
                                    },
                                  ],
                                  "name": "prefix",
                                },
                                {
                                  "children": [],
                                  "name": "name",
                                  "value": "A",
                                },
                              ],
                              "name": "infix",
                            },
                            {
                              "children": [],
                              "name": "name",
                              "value": "j",
                            },
                          ],
                          "name": "infix",
                        },
                        {
                          "children": [],
                          "name": "name",
                          "value": "i",
                        },
                      ],
                      "name": "infix",
                    },
                    {
                      "children": [],
                      "name": "number",
                      "value": 127,
                    },
                  ],
                  "name": "infix",
                },
              ],
              "name": "group",
              "value": "parens",
            },
          ],
          "name": "infix",
        },
      ],
      "name": "infix",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "/",
        },
        {
          "children": [],
          "name": "number",
          "value": 1,
        },
        {
          "children": [],
          "name": "number",
          "value": 1,
        },
      ],
      "name": "infix",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > produces correct tree for '++i' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "prefixIncrement",
    },
    {
      "children": [],
      "name": "name",
      "value": "i",
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > produces correct tree for '-(a+b)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "negate",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "+",
            },
            {
              "children": [],
              "name": "name",
              "value": "a",
            },
            {
              "children": [],
              "name": "name",
              "value": "b",
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > produces correct tree for '--i' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "prefixDecrement",
    },
    {
      "children": [],
      "name": "name",
      "value": "i",
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > arithmetics > produces correct tree for '1 + 2^3 * 4' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [],
      "name": "number",
      "value": 1,
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "*",
        },
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "^",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
        {
          "children": [],
          "name": "number",
          "value": 4,
        },
      ],
      "name": "infix",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > arithmetics > subtract 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "-",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > data structures > produces correct tree for '(-(2+7)/A,j, i, 127.0 )' 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": ",",
        },
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "group",
                  "value": ",",
                },
                {
                  "children": [
                    {
                      "children": [],
                      "name": "group",
                      "value": "/",
                    },
                    {
                      "children": [
                        {
                          "children": [],
                          "name": "group",
                          "value": "negate",
                        },
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [],
                                  "name": "group",
                                  "value": "+",
                                },
                                {
                                  "children": [],
                                  "name": "number",
                                  "value": 2,
                                },
                                {
                                  "children": [],
                                  "name": "number",
                                  "value": 7,
                                },
                              ],
                              "name": "infix",
                            },
                          ],
                          "name": "group",
                          "value": "parens",
                        },
                      ],
                      "name": "prefix",
                    },
                    {
                      "children": [],
                      "name": "name",
                      "value": "A",
                    },
                  ],
                  "name": "infix",
                },
                {
                  "children": [],
                  "name": "name",
                  "value": "j",
                },
              ],
              "name": "infix",
            },
            {
              "children": [],
              "name": "name",
              "value": "i",
            },
          ],
          "name": "infix",
        },
        {
          "children": [],
          "name": "number",
          "value": 127,
        },
      ],
      "name": "infix",
    },
  ],
  "name": "group",
  "value": "parens",
}
`;

exports[`expressions > data structures > tuple 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": ",",
    },
    {
      "children": [],
      "name": "number",
      "value": 1,
    },
    {
      "children": [],
      "name": "number",
      "value": 2,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > data structures > unit 1`] = `
{
  "children": [],
  "name": "group",
  "value": "parens",
}
`;

exports[`expressions > fixity expressions > group 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "number",
      "value": 1,
    },
  ],
  "name": "group",
  "value": "parens",
}
`;

exports[`expressions > fixity expressions > infix 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 456,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > fixity expressions > mixfix 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "name": "number",
          "value": 456,
        },
      ],
      "name": "group",
      "value": "inRange",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
    {
      "children": [],
      "name": "number",
      "value": 789,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > fixity expressions > name 1`] = `
{
  "children": [],
  "name": "name",
  "value": "name",
}
`;

exports[`expressions > fixity expressions > postfix 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "postfixDecrement",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "postfix",
}
`;

exports[`expressions > fixity expressions > prefix 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "name",
      "value": "+",
    },
    {
      "children": [],
      "name": "number",
      "value": 123,
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > function expressions > application > function call 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [],
      "name": "name",
      "value": "f",
    },
    {
      "children": [],
      "name": "name",
      "value": "x",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > function call multiple args 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "application",
        },
        {
          "children": [],
          "name": "name",
          "value": "f",
        },
        {
          "children": [],
          "name": "name",
          "value": "x",
        },
      ],
      "name": "infix",
    },
    {
      "children": [],
      "name": "name",
      "value": "y",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for '(send 1 2)(2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "application",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "group",
                  "value": "application",
                },
                {
                  "children": [],
                  "name": "name",
                  "value": "send",
                },
                {
                  "children": [],
                  "name": "number",
                  "value": 1,
                },
              ],
              "name": "infix",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for '(send 1)(2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "application",
            },
            {
              "children": [],
              "name": "name",
              "value": "send",
            },
            {
              "children": [],
              "name": "number",
              "value": 1,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for '(send)(2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [],
          "name": "name",
          "value": "send",
        },
      ],
      "name": "group",
      "value": "parens",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'a + send (2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [],
      "name": "name",
      "value": "a",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "application",
        },
        {
          "children": [],
          "name": "name",
          "value": "send",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "name": "group",
                  "value": ",",
                },
                {
                  "children": [],
                  "name": "number",
                  "value": 2,
                },
                {
                  "children": [],
                  "name": "number",
                  "value": 3,
                },
              ],
              "name": "infix",
            },
          ],
          "name": "group",
          "value": "parens",
        },
      ],
      "name": "infix",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'a + send 1 + 2' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "+",
        },
        {
          "children": [],
          "name": "name",
          "value": "a",
        },
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "application",
            },
            {
              "children": [],
              "name": "name",
              "value": "send",
            },
            {
              "children": [],
              "name": "number",
              "value": 1,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "infix",
    },
    {
      "children": [],
      "name": "number",
      "value": 2,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'send 1 (2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "application",
        },
        {
          "children": [],
          "name": "name",
          "value": "send",
        },
        {
          "children": [],
          "name": "number",
          "value": 1,
        },
      ],
      "name": "infix",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'send 1 + 2' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "+",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "application",
        },
        {
          "children": [],
          "name": "name",
          "value": "send",
        },
        {
          "children": [],
          "name": "number",
          "value": 1,
        },
      ],
      "name": "infix",
    },
    {
      "children": [],
      "name": "number",
      "value": 2,
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'send a (2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "application",
        },
        {
          "children": [],
          "name": "name",
          "value": "send",
        },
        {
          "children": [],
          "name": "name",
          "value": "a",
        },
      ],
      "name": "infix",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'send((1+2), 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "send",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "children": [],
                      "name": "group",
                      "value": "+",
                    },
                    {
                      "children": [],
                      "name": "number",
                      "value": 1,
                    },
                    {
                      "children": [],
                      "name": "number",
                      "value": 2,
                    },
                  ],
                  "name": "infix",
                },
              ],
              "name": "group",
              "value": "parens",
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > function expressions > application > produces correct tree for 'send(2, 3)' 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "application",
    },
    {
      "children": [],
      "name": "name",
      "value": "send",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": ",",
            },
            {
              "children": [],
              "name": "number",
              "value": 2,
            },
            {
              "children": [],
              "name": "number",
              "value": 3,
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "parens",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > function expressions > funciton multiple params 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "group",
              "value": "application",
            },
            {
              "children": [],
              "name": "name",
              "value": "x",
            },
            {
              "children": [],
              "name": "name",
              "value": "y",
            },
          ],
          "name": "infix",
        },
      ],
      "name": "group",
      "value": "fn",
    },
    {
      "children": [
        {
          "children": [],
          "name": "group",
          "value": "+",
        },
        {
          "children": [],
          "name": "name",
          "value": "x",
        },
        {
          "children": [],
          "name": "name",
          "value": "y",
        },
      ],
      "name": "infix",
    },
  ],
  "name": "prefix",
}
`;

exports[`expressions > function expressions > function 1`] = `
{
  "children": [
    {
      "children": [],
      "name": "group",
      "value": "->",
    },
    {
      "children": [],
      "name": "name",
      "value": "x",
    },
    {
      "children": [],
      "name": "name",
      "value": "x",
    },
  ],
  "name": "infix",
}
`;

exports[`expressions > values > number 1`] = `
{
  "children": [],
  "name": "number",
  "value": 123,
}
`;

exports[`expressions > values > string 1`] = `
{
  "children": [],
  "name": "string",
  "value": "string",
}
`;

exports[`values > number 1`] = `
{
  "children": [],
  "name": "number",
  "value": 123,
}
`;

exports[`values > string 1`] = `
{
  "children": [],
  "name": "string",
  "value": "string",
}
`;
